RoaringBitmap/src/main/java/org/roaringbitmap/FastAggregation.java

```diff
@@ -355,6 +393,94 @@ public static RoaringBitmap workShyAnd(long[] buffer, RoaringBitmap... bitmaps)
     return new RoaringBitmap(array);
   }
 
+  private static int workShyAndCardinality(RoaringBitmap... bitmaps) {
+    long[] words = new long[1024];
+    RoaringBitmap first = bitmaps[0];
+    for (int i = 0; i < first.highLowContainer.size; ++i) {
+      char key = first.highLowContainer.keys[i];
+      words[key >>> 6] |= 1L << key;
+    }
+    int numContainers = first.highLowContainer.size;
+    for (int i = 1; i < bitmaps.length && numContainers > 0; ++i) {
+      numContainers = Util.intersectArrayIntoBitmap(words,
+          bitmaps[i].highLowContainer.keys, bitmaps[i].highLowContainer.size);
+    }
+    if (numContainers == 0) {
+      return 0;
+    }
+    Container[][] containers = new Container[numContainers][bitmaps.length];
+    for (int i = 0; i < bitmaps.length; ++i) {
+      RoaringBitmap bitmap = bitmaps[i];
+      int position = 0;
+      for (int j = 0; j < bitmap.highLowContainer.size; ++j) {
+        char key = bitmap.highLowContainer.keys[j];
+        if ((words[key >>> 6] & (1L << key)) != 0) {
+          containers[position++][i] = bitmap.highLowContainer.values[j];
+        }
+      }
+    }
+
+    int cardinality = 0;
+    for (int i = 0; i < numContainers; ++i) {
+      Container[] slice = containers[i];
+      Arrays.fill(words, -1L);
+      Container tmp = new BitmapContainer(words, -1);
+      for (Container container : slice) {
+        Container and = tmp.iand(container);
+        if (and != tmp) {
+          tmp = and;
+        }
+      }
+      tmp = tmp.repairAfterLazy();
+      cardinality += tmp.getCardinality();
+    }
+    return cardinality;
+  }
+
+  private static int horizontalOrCardinality(RoaringBitmap... bitmaps) {
+    long[] words = new long[1024];
+    int minKey = Character.MAX_VALUE;
+    int maxKey = Character.MIN_VALUE;
```

Comment:
If you were to allocate the bitset ("words") after having figured out the minKey and maxKey, you could reduce the memory overhead. Also 'words' could possibly be implemented using a standard Bitset.

Reply:
It's needed for the intersection at the bottom, so this wouldn't save memory in the end.